### PR TITLE
[llhd] Refactor struct Signal with class Signal 

### PR DESCRIPTION
**Pull request by DeepFlyingSky**
Refactor struct Signal with the class type, add with basic class methods.

updateWhenChanged method can speed up compare to std::memcpy method.

### CODE REVIEW

File: include/circt/Dialect/LLHD/Simulator/State.h [hunk 1]
```diff
@@ -79,12 +83,98 @@ struct Signal {
...
 
   /// Returns true if the owner name is lexically smaller than rhs's owner, or
   /// the name is lexically smaller than rhs's name, in case they share the same
   /// owner.
-  bool operator<(const Signal &rhs) const;
+  bool operator<(const Signal &rhs) const {
+    return (owner < rhs.owner || (owner == rhs.owner && name < rhs.name));
+  }
+
+  bool isOwner(std::string rhs) const { return owner == rhs; };
```
Comment (maerhart) [inline include/circt/Dialect/LLHD/Simulator/State.h:99]: Why not pass the string by (const) reference?

File: lib/Dialect/LLHD/Simulator/Trace.cpp [hunk 2]
```diff
@@ -24,15 +24,10 @@ Trace::Trace(std::unique_ptr<State> const &state, llvm::raw_ostream &out,
     : out(out), state(state), mode(mode) {
   auto root = state->root;
   for (auto &sig : state->signals) {
-    if (mode != TraceMode::Full && mode != TraceMode::Merged &&
-        sig.owner != root) {
-      isTraced.push_back(false);
-    } else if (mode == TraceMode::NamedOnly &&
-               std::regex_match(sig.name, std::regex("(sig)?[0-9]*"))) {
```
Comment (maerhart) [inline lib/Dialect/LLHD/Simulator/Trace.cpp:31]: I think the `#include <regex>` at the top of the file can be removed now?

**Review by maerhart — APPROVED** [review]
Nice work. Thanks for pushing on this!

**Comment by DeepFlyingSky** [comment]
Thanks for review @maerhart .

**Pull request merged**